File: articles/selfhood-agency-and-personal-identity-in-personality-psychology/r/analyze_selfhood_agency_identity.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_selfhood_agency_identity.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

data$continuity_level <- ifelse(
  data$temporal_self_continuity > median(data$temporal_self_continuity),
  "higher_continuity",
  "lower_continuity"
)
data$agency_level <- ifelse(
  data$situated_agency_index > median(data$situated_agency_index),
  "higher_agency",
  "lower_agency"
)
data$identity_profile <- paste(data$continuity_level, data$agency_level, sep = "_")
data$high_constraint_low_agency <- data$external_constraint > median(data$external_constraint) &
  data$situated_agency_index < median(data$situated_agency_index)
data$low_continuity_low_integration <- data$temporal_self_continuity < median(data$temporal_self_continuity) &
  data$identity_integration < median(data$identity_integration)

context_summary <- aggregate(
  data[, c("temporal_self_continuity", "agency_index", "situated_agency_index", "social_recognition", "external_constraint", "identity_integration", "well_being")],
  by = list(identity_context = data$identity_context),
  FUN = mean
)
write.csv(context_summary, file.path(outputs, "r_identity_context_summary.csv"), row.names = FALSE)

profile_summary <- aggregate(
  data[, c("temporal_self_continuity", "situated_agency_index", "social_recognition", "external_constraint", "identity_integration", "well_being")],
  by = list(identity_profile = data$identity_profile),
  FUN = mean
)
write.csv(profile_summary, file.path(outputs, "r_identity_profile_summary.csv"), row.names = FALSE)

corr_cols <- c(
  "past_self", "present_self", "future_self", "intentional_clarity", "action_ownership",
  "self_efficacy", "external_constraint", "social_recognition", "value_commitment_gap",
  "identity_integration", "well_being", "past_present_gap", "present_future_gap",
  "temporal_self_continuity", "agency_index", "situated_agency_index", "identity_alignment"
)
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_selfhood_agency_identity_correlations.csv"))

model_identity <- lm(
  identity_integration ~ temporal_self_continuity + situated_agency_index +
    social_recognition + external_constraint + value_commitment_gap,
  data = data
)

model_wellbeing <- lm(
  well_being ~ identity_integration + situated_agency_index +
    temporal_self_continuity + social_recognition + external_constraint,
  data = data
)

model_agency <- lm(
  agency_index ~ intentional_clarity + action_ownership + self_efficacy +
    social_recognition + external_constraint,
  data = data
)

capture.output(summary(model_identity), file = file.path(outputs, "r_identity_integration_model_summary.txt"))
capture.output(summary(model_wellbeing), file = file.path(outputs, "r_wellbeing_model_summary.txt"))
capture.output(summary(model_agency), file = file.path(outputs, "r_agency_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_selfhood_agency_identity.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")

File: articles/selfhood-agency-and-personal-identity-in-personality-psychology/python/analyze_selfhood_agency_identity.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_selfhood_agency_identity.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

BASE = [
    "past_self",
    "present_self",
    "future_self",
    "intentional_clarity",
    "action_ownership",
    "self_efficacy",
    "external_constraint",
    "social_recognition",
    "value_commitment_gap",
    "identity_integration",
    "well_being",
]
DERIVED = [
    "past_present_gap",
    "present_future_gap",
    "temporal_self_continuity",
    "agency_index",
    "situated_agency_index",
    "identity_alignment",
]

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    return dict(zip(["intercept", *xs], map(float, beta)))

df = pd.read_csv(DATA)
required = {"person_id", "identity_context", *BASE, *DERIVED}
missing = required - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

df["continuity_level"] = np.where(
    df["temporal_self_continuity"] > df["temporal_self_continuity"].median(),
    "higher_continuity",
    "lower_continuity",
)
df["agency_level"] = np.where(
    df["situated_agency_index"] > df["situated_agency_index"].median(),
    "higher_agency",
    "lower_agency",
)
df["identity_profile"] = df["continuity_level"] + "_" + df["agency_level"]
df["high_constraint_low_agency"] = (
    (df["external_constraint"] > df["external_constraint"].median())
    & (df["situated_agency_index"] < df["situated_agency_index"].median())
)
df["low_continuity_low_integration"] = (
    (df["temporal_self_continuity"] < df["temporal_self_continuity"].median())
    & (df["identity_integration"] < df["identity_integration"].median())
)

context_summary = (
    df.groupby("identity_context")
    .agg(
        n=("person_id", "count"),
        temporal_self_continuity_mean=("temporal_self_continuity", "mean"),
        agency_index_mean=("agency_index", "mean"),
        situated_agency_mean=("situated_agency_index", "mean"),
        social_recognition_mean=("social_recognition", "mean"),
        external_constraint_mean=("external_constraint", "mean"),
        identity_integration_mean=("identity_integration", "mean"),
        well_being_mean=("well_being", "mean"),
    )
    .reset_index()
)

profile_summary = (
    df.groupby("identity_profile")
    .agg(
        n=("person_id", "count"),
        temporal_self_continuity_mean=("temporal_self_continuity", "mean"),
        situated_agency_mean=("situated_agency_index", "mean"),
        social_recognition_mean=("social_recognition", "mean"),
        external_constraint_mean=("external_constraint", "mean"),
        identity_integration_mean=("identity_integration", "mean"),
        well_being_mean=("well_being", "mean"),
    )
    .reset_index()
)

disjunction_summary = pd.DataFrame(
    {
        "pattern": ["high_constraint_low_agency", "low_continuity_low_integration"],
        "n": [int(df["high_constraint_low_agency"].sum()), int(df["low_continuity_low_integration"].sum())],
        "proportion": [float(df["high_constraint_low_agency"].mean()), float(df["low_continuity_low_integration"].mean())],
    }
)

corr_cols = BASE + DERIVED
models = {
    "identity_integration": ols(df, "identity_integration", ["temporal_self_continuity", "situated_agency_index", "social_recognition", "external_constraint", "value_commitment_gap"]),
    "well_being": ols(df, "well_being", ["identity_integration", "situated_agency_index", "temporal_self_continuity", "social_recognition", "external_constraint"]),
    "agency_index": ols(df, "agency_index", ["intentional_clarity", "action_ownership", "self_efficacy", "social_recognition", "external_constraint"]),
    "note": "Synthetic demonstration only; not clinical assessment, identity evaluation, personality testing, workplace screening, legal evaluation, or individual prediction.",
}

context_summary.to_csv(OUT / "python_identity_context_summary.csv", index=False)
profile_summary.to_csv(OUT / "python_identity_profile_summary.csv", index=False)
disjunction_summary.to_csv(OUT / "python_identity_disjunction_summary.csv", index=False)
df[corr_cols].corr().to_csv(OUT / "python_selfhood_agency_identity_correlations.csv")
(OUT / "python_selfhood_agency_identity_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_selfhood_agency_identity.csv", index=False)
print(f"Wrote outputs to {OUT}")
